### PR TITLE
Include stock item purchase price in pricing cache

### PR DESCRIPTION
**Pull request by SchrodingersGat**
Ref: https://github.com/inventree/InvenTree/issues/4262

This PR allows purchase price data from individual stock items to be included when calculating (and caching) part pricing data. This is of particular interest when stock items are entered manually, without a referenced purchase order.

### Changes

- Adds new cached pricing fields (`stock_item_cost_min` and `stock_item_cost_max`)
- Allows these data to be used for overall pricing calculations
- Pricing is automatically updated whenever stock items are created / edited / deleted
- Unit testing

Closes https://github.com/inventree/InvenTree/issues/4262

<a href="https://gitpod.io/#https://github.com/inventree/InvenTree/pull/4292"></a>

@simonkuehling @cabottech @matmair

### CODE REVIEW

File: InvenTree/part/test_pricing.py [hunk 1]
```diff
@@ -234,6 +235,53 @@ def test_internal_pricing(self):
...
+        self.assertIsNone(pricing.overall_min)
+        self.assertIsNone(pricing.overall_max)
+
+        # Turn on stock pricing
+        common.models.InvenTreeSetting.set_setting('PRICING_USE_STOCK_PRICING', True, None)
+
+        pricing.update_pricing()
+
+        self.assertEqual(pricing.overall_min, Money(1.176471, 'USD'))
+        self.assertEqual(pricing.overall_max, Money(6.666667, 'USD'))
```
Comment (matmair) [inline InvenTree/part/test_pricing.py:283]: Suggestion: Maybe an explanation of where the numbers come from would be helpful for understanding the test in the future

File: InvenTree/stock/models.py [hunk 2]
```diff
@@ -2001,6 +2005,21 @@ def after_save_stock_item(sender, instance: StockItem, created, **kwargs):
...
 
+        # Determine if pricing should be updated for the parent Part instance
+        pricing = instance.part.pricing
+
+        # Check if the stock_item pricing will change
+        stock_min = pricing.stock_item_cost_min
+        stock_max = pricing.stock_item_cost_max
+
+        # Recalculate stock item pricing (but do not save)
+        pricing.update_stock_item_cost(save=False)
```
Comment (matmair) [inline InvenTree/stock/models.py:2016]: Question: How database extensive is this?

**Review by matmair — APPROVED** [review]
Looks good!

File: InvenTree/part/models.py [hunk 3]
```diff
@@ -2516,6 +2517,45 @@ def update_purchase_cost(self, save=True):
...
+
+        min_item_cost = None
+        max_item_cost = None
+
+        for item in items:
+            cost = self.convert(item.purchase_price)
+
+            # Skip if the cost could not be converted (for some reason)
+            if cost is None:
+                continue
```
[the remark this answers is not in the document]
Comment (SchrodingersGat) [reply InvenTree/part/models.py:2545]: Most likely is that the correct exchange rate is missing (that's the only error I've seen to date)

File: InvenTree/part/templates/part/prices.html [hunk 4]
```diff
@@ -60,12 +58,21 @@ <h4>{% trans "Pricing Overview" %}</h4>
...
                     </tr>
+                    {% settings_value "PRICING_USE_STOCK_PRICING" as use_stock_pricing %}
+                    {% if use_stock_pricing %}
+                    <tr>
+                        <td>
+                            <span class='fas fa-boxes'></span>
+                        </td>
+                        <td>{% trans "Stock Item Pricing" %}</td>
+                        <td>{% include "price_data.html" with price=pricing.stock_item_cost_min %}</td>
+                        <td>{% include "price_data.html" with price=pricing.stock_item_cost_max %}</td>
```
[the remark this answers is not in the document]
Comment (SchrodingersGat) [reply InvenTree/part/templates/part/prices.html:73]: @matmair good point. Probably could just roll them together.

File: InvenTree/part/models.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (matmair) [reply InvenTree/part/models.py]: ok that seems fine to not log

**Pull request merged**